### ISSUE TITLE
Inconsistent handling of names/labels/identifiers

### ISSUE THREAD

**Issue by danielwe** [mabuchilab/QNET#53]
When creating a circuit component, such as a ``Phase``, the first positional argument is the name used to identify the component. The current handling of this argument is a little confusing. The code is (from `component.py`):
```python
    def __init__(self, name, **kwargs):
        self._name = str(name)
        if not self._rx_name.match(name):
            raise ValueError("name '%s' does not match pattern '%s'"
                             % (self.name, self._rx_name.pattern))n
        ...
```
The first line implies that we are lenient about the type of ``name``, and will convert it to a string if at all possible. However, the regex matching is performed on the original parameter and will raise an error if ``name`` was not a string initially. I'm wondering what the desired behavior here actually is?

I think I would err on the stricter side because the lenient behavior exposes some gotchas. For example, say you need a phase shift by some symbolic angle phi, and you go
```python
phi = sympy.symbols('phi')
phase = Phase(phi)
#phase = Phase('phi', phi=phi)  # This would have been the correct way
```
Currently, this raises an error. The more lenient behavior would set ``phase.name`` to ``str(phi)``, which is just ``'phi'``. As a user, you'd be forgiven for making the intuitive guess that it would also associate the symbolic variable ``phi`` with the phase angle, i.e., that ``phase.phi is phi`` would be True, but this is not the case, since only the name parameter was passed. Adding to the confusion, ``phase.phi`` would default to the class variable ``phi = sympy.symbols('phi', real=True)``, and hence be indistinguishable from ``phi`` in any expression. However, since they were instantiated with different parameters (none vs. ``real=True``), they are distinct objects, something you'll only realize once you begin substituting things for ``phi``.

If you instead called your angle, e.g., ``'theta'``, a whole different slew of confusions would arise.

TLDR: being a little strict may prevent a lot of confusion. Would you agree that this is the way to go?

I tried to look for precedence elsewhere in the code. I looked at all cases where a name/label/identifier is checked against a regex pattern in ``__init__``, and it turns out that this is done in a few different ways. The candidates are:
* ``KetSymbol`` in `state_algebra.py` (super strict, explicitly enforces ``isinstance(label, str)``),
* ``CircuitSymbol`` in `circuit_algebra.py` (lenient, uses ``str(name)``),
* ``LocalOperator`` in `operator_algebra.py` (strict, calls regex ``match(identifier)`` directly),
* ``OperatorSymbol`` in `operator_algebra.py` (lenient, uses ``str(identifier)``),
* ``SuperOperatorSymbol`` in `super_operator_algebra.py` (strict, calls regex ``match(label)`` directly),
* ``LocalSpace`` in `hilbert_space_algebra.py` (lenient, uses``str(label)``),
* ``Component`` in `component.py` (the case referred to above: confused, calls ``str(name)`` but doesnt use the result when calling regex ``match(name)``).

It might be a good idea to decide on a common behavior for all of these. I can update them once we decide which.

**Comment by goerz** [comment]
The example a bug. That should be

    name = str(name)

and then

    self._name = name

farther below (after matching).

I also completely agree that we should have a consistent handling. I'll have to think a little bit on whether we want the more lenient or the more strict version. I think the main idea behind "lenient" was allowing integer labels in some cases.

(I  think I might be leaning towards strict, `assert isinstance(name, str)`)

**Comment by ngedwin98** [comment]
+1 for the listing all the gotchas I've fallen into -- and then some. As things currently stand, I'd also prefer the stricter behavior. However, I think the confusion caused by this handling issue is due in no small part to problematic semantic choices in the `circuit_components` submodule.

For one, variable names for `Component` parameters ought to be symbol-agnostic, e.g., `phase` or `angle` as opposed to `phi`. The symbol is merely a placeholder for a value, while the parameter itself is a property of the component. Similarly, for a cavity-based component, I wouldn't call the detuning `Delta`, but rather `detuning`, and so on.

To furthermore have the value of the parameter default to a symbol-convention-dependent value like `sympy.symbols('phi', real=True)` only exacerbates the problem. It would be more intuitive to choose a neutral default value, such as one which makes the component behave the same way as `cid(self.cdim)` (in this case, 0), perhaps with the value even defaulting to `None` in the event that this is not possible (as long as this doesn't cause some nasty behavior elsewhere).

Incidentally, a feature in which `Phase(phi)` produces a `Phase` with a default name of `"Phase"` and a phase angle of `phi` is actually quite attractive. If the semantic issues can be solved, a more radical change in the handling could relegate the name argument to last, defaulting to the class name (which is often what I end up naming the component anyway). Given the current implementation via the `_parameters` attribute, I imagine it might be tricky to have this, but having such an admittedly intuitive behavior be the default might also help alleviate these pitfalls.

**Comment by danielwe** [comment]
I agree that the parameter naming and default values are questionable. And what is the name used for anyway? (Exposing my limited practical experience with the package here...)

**Comment by danielwe** [comment]
By the way, wouldn't it be better to raise a `TypeError` rather than using an `assert` statement? And we could, of course, keep the support for integer labels as well.

**Comment by goerz** [comment]
Yeah, raising a `TypeError` is better than `assert`. Also just letting the `match` throw an exception might be fine, depending on how clear the error message is.

A little bit of background on the three similar terms name/label/identifier: `name` is primarily used in `circuit_algebra` to give a label to a component to be used when rendering a circuit-algebra formula to TeX or into a diagram. These often include a word, like `'Phase_1'`. The variable `identifier` is primarily used in `operator_algebra`. It's also a "symbol" used for the operator when rendering a formula. This is often a single letter (with possible subscript). Technically, though, `name` and `identifier` are basically the same thing, and we could consider using only one term (`name`?). The `label` used e.g. in the `state_algebra` is slightly different: it's usually the label of an eigenstate.  Internally, this is always a string, but very often it's just "0", "1", "2", ...  (and one might want to allow instantiating with an integer that is the converted to a string -- the motivation for having the `str` conversion in the code). I'm also considering allowing labels to be sympy symbols in the future, in order to allow for symbolic sums (`\sum_i \ket{i}`). This isn't fully though out yet, though.

I completely agree with @ngedwin98 that the parameter names in the components should be symbol-agnostic (`phase` or `angle` instead of `phi`).

As to avoid gotchas: Since we've switched to Python3 only, keyword-only arguments would be an option. I generally quite like them: they force the user to write more verbose code, but they really make sure there are not surprises. The constructor for the `Component` class could use one of the following signatures

    def __init__(self, *, name, **kwargs):

    def __init__(self, **kwargs):

where the latter one assumes that there's going to be a default `name` for a given component (e.g. `Phase`, as you're proposing), and `name` is still in `kwargs` to allow to overwrite it. Not completely sure yet if this is a good idea.  The first option would be the way to go if we decide that default names don't really work. That would prevent the user from accidentally writing `phase = Phase(phi)`. They'd be forced to write `phase = Phase(name='Phase', phi=phi)`. Verbose, but hard to get wrong.

**Comment by goerz** [comment]
Pretty much the entire "popular demand" behavior is now implemented:

* Components now take keyword arguments only (i.e., maximum verbosity/safety), with sensible defaults.
* I wasn't really on board with default values that make the components into identities. Using symbols as defaults makes sense to me (especially because they have the correct algebraic properties).
* The *name* (now *label*) is primarily used for printing, and is optional now (there's a sensible default; not necessarily the class name, but e.g. `BS` for `Beamsplitter`). The one caveat is that if different physical components are not distinguished by their label, `connect` becomes more cumbersome: components are no longer uniquely identifiable, and must be manually identified by their index in the list of components. It's a small price to pay.
* Components are also subclasses of `CircuitSymbol`, which guarantees concise printing despite the verbose keyword-only instantiation.
* The argument names are now symbol-agnostic, e.g. `Phase` has an argument `phase` instead of `phi`
